### ISSUE TITLE
make module context-aware

### ISSUE THREAD

**Issue by mac-g** [Automattic/node-canvas#1394]
## Issue or Feature
I've been experimenting with worker threads in node.js recently and noticed that the canvas module does not appear to be context-aware. In other words, the module will load/instantiate once, but as soon as another thread attempts to require canvas for it's own use, the following error is thrown:
```sh
Error: Module did not self-register.
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:840:18)
    at Module.load (internal/modules/cjs/loader.js:666:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:606:12)
    at Function.Module._load (internal/modules/cjs/loader.js:598:3)
    at Module.require (internal/modules/cjs/loader.js:705:19)
    at require (internal/modules/cjs/helpers.js:14:16)
    at Object.<anonymous> (~/node_modules/canvas/lib/bindings.js:3:18)
    at Module._compile (internal/modules/cjs/loader.js:799:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:810:10)
    at Module.load (internal/modules/cjs/loader.js:666:32)
```
## Steps to Reproduce

In main thread...
```js
const {Worker} = require('worker_threads');
var worker1 = new Worker('./child-thread.js'); 
var worker2 = new Worker('./child-thread.js');
// etc.
```
In child-thread.js...
```js
var Canvas = require('canvas');
var canvas = Canvas.createCanvas(200, 200);
var ctx = canvas.getContext('2d');
// etc.
```
## References
**Node.js documentation:**
[https://nodejs.org/api/addons.html#addons_context_aware_addons](https://nodejs.org/api/addons.html#addons_context_aware_addons)
**Similar issue:**
[https://github.com/schroffl/node-lzo/pull/11](https://github.com/schroffl/node-lzo/pull/11)

## Environment
* node-canvas version **2.4.1**
* Mac OS **10.14.3**
* node version **11.12.1**

**Comment by LinusU** [comment]
PRs welcome 👍

**Comment by jgoralcz** [comment]
This would be an incredible addition to node-canvas. Not sure where to fully dive in and modify the code though 🤔

**Comment by ruiwei** [comment]
I'm having the same issue. Any solution yet?

**Comment by a-koka** [comment]
Nope. Need to fork processes and can't use worker threads.

**Comment by siddarthareddyt** [comment]
https://github.com/Automattic/node-canvas/pull/1409 doesn't seem to fix this issue. Any plans to provide support for `worker_threads`?

**Comment by a-dpq** [comment]
I'm having the same issue. I'd love to be able to use node-canvas in multiple worker threads to improve performance. Anyone want to tackle this? Seems like it should be straightforward to solve:

[https://github.com/nodejs/node/blob/master/doc/api/addons.md#context-aware-addons](https://github.com/nodejs/node/blob/master/doc/api/addons.md#context-aware-addons)

**Comment by mistval** [comment]
This would be very useful.

**Comment by lateiskate** [comment]
I was redirected here. I'd really love for this to be a thing. Using child processes is very expensive unlike worker threads

**Comment by aya1337** [comment]
This is needed

**Comment by mac-g** [comment]
Looks like I'll have a chance to _maybe_ take a stab at this in a week while I'm on leave. No promises. Has anyone else had a chance to take a look at it? If so, I'd be great to avoid hours of re-discovery.

**Comment by lostrepo** [comment]
Maybe this will help.

I'm total noob in C++
But I use something like that to use h264 encoder in worker_threads
```c++
#include "v8.h"
#include <node.h>
#include <node_buffer.h>

using namespace v8;
```

Per NodeJS instance addon data
```c++
class AddonData {
	public:
	explicit AddonData(Isolate* isolate):
	wh(0) {
		// Ensure this per-addon-instance data is deleted at environment cleanup
		node::AddEnvironmentCleanupHook(isolate, DeleteInstance, this);
	}
	
	
	// Per-addon data
	int wh;
	// ... omitted
	// Node Buffer Object for YUV420 source
	MaybeLocal<Object> ab;
	// Node Buffer Object for h264 bitstream
	MaybeLocal<Object> bsiab;
	
	
	int createEncoder(){
		int rv	=	WelsCreateSVCEncoder(&encoder_);
	
		return encoder_ != NULL ? rv : 1;
	}

	void clean() {
		if (encoder_) {
			encoder_->Uninitialize();
			WelsDestroySVCEncoder(encoder_);
		}
	}
	
	
	static void abDeleteCb(char* data, void* hint) {
		delete reinterpret_cast<MaybeLocal<Object>*>(hint);
	}
	
	static void noDeleteCb(char* data, void* hint) {
		// do nothing
	}
	
	static void DeleteInstance(void* data) {
		AddonData* ndata	=	reinterpret_cast<AddonData*>(data);
		ndata->clean();
		
		delete ndata;
	}
};
```

Callback that accesses per NodeJS instance addon data
```c++
void encode(const FunctionCallbackInfo<Value>& info) {
	// Retrieve the per-addon-instance data
	AddonData* data	=	reinterpret_cast<AddonData*>(info.Data().As<External>()->Value());
	
	// encode YUV420 from buf
	data->encode();
	// construct h264 bitstream from bitstream info
	data->parseBitstreamInfo();
	
	data->bsiab	=	node::Buffer::New(info.GetIsolate(), (char*) data->bsi.data(), (size_t) data->bsiSize, AddonData::abDeleteCb, (void*) &data->bsiab);
	
	info.GetReturnValue().Set(data->bsiab.ToLocalChecked());
}
```

Per NodeJS instance module initializer (another method of initialization can be used too but this one was used in another addon that works with worker_threads and I'm total C++ noob)
```c++
extern "C" NODE_MODULE_EXPORT void
NODE_MODULE_INITIALIZER(Local<Object> exports, Local<Value> module, Local<Context> context) {
	Isolate* isolate	=	context->GetIsolate();
	
	
	// Create a new instance of `AddonData` for this instance of the addon and
	// tie its life cycle to that of the Node.js environment
	AddonData* data	=	new AddonData(isolate);
	
	// create encoder instance
	data->createEncoder();
	
	
	// Wrap the data in a `v8::External` so we can pass it to the method we expose
	Local<External> external = External::New(isolate, data);
	
	
	// ... omitted
	
	exports->Set(
		context,
		String::NewFromUtf8(isolate, "encode", NewStringType::kNormal).ToLocalChecked(),
		FunctionTemplate::New(isolate, getab, external)->GetFunction(context).ToLocalChecked()
	).FromJust();
}
```

**Comment by maxreisner** [comment]
I tried getting this to work with no success... If anyone who's more c++ oriented than me (and I'm not) would want some help fixing this, I'd be happy to help!

**Comment by Mjtlittle** [comment]
I hope i'm not leading anyone astray, as I have no clue how the node API bindings work with creating a [context-aware module](https://github.com/nodejs/node/blob/master/doc/api/addons.md#context-aware-addons); but I found a module called 'node-lzo' that has made changes to adapt their library, that I don't see present here.

https://github.com/schroffl/node-lzo/pull/11

specifically this...

```cpp
NODE_MODULE_INIT(/* exports, module, context */) {
    Init(exports, context);
}
```

**Comment by enricodeleo** [comment]
+1

**Comment by dapriett** [comment]
+1

**Comment by lovanwubing** [comment]
+1

**Comment by dolcalmi** [comment]
> I hope i'm not leading anyone astray, as I have no clue how the node API bindings work with creating a [context-aware module](https://github.com/nodejs/node/blob/master/doc/api/addons.md#context-aware-addons); but I found a module called 'node-lzo' that has made changes to adapt their library, that I don't see present here.
> 
> [schroffl/node-lzo#11](https://github.com/schroffl/node-lzo/pull/11)
> 
> specifically this replacement...
> 
> ```c++
> NODE_MODULE_INIT(/* exports, module, context */) {
>     Init(exports, context);
> }
> ```

It works... I had to recompile it with the next change to make it work with worker threads
```
// NODE_MODULE(canvas, init);
NODE_MODULE_INIT() {
    init(exports);
}
```

**Comment by enricodeleo** [comment]
@dolcalmi @Mjtlittle can you please submit a pull request? I think we all would take advantage of workers support

**Comment by danielyaa5** [comment]
+1 @dolcalmi @Mjtlittle please!

**Comment by danielyaa5** [comment]
I tried this actually and I dont think its enough.

After I compile and run it in worker thread I get this error 

```
TypeError: Canvas expected
```

I'm trying to understand the proper way to do this but I have no idea

**Comment by dteh** [comment]
@danielyaa5 I can confirm recompiling with the `NODE_MODULE_INIT` macro worked for me. canvas.createCanvas(200,200) returned [Canvas 200x200].

**Comment by yeldarby** [comment]
@dolcalmi or @dteh -- is there a fork that has this change available somewhere?

**Comment by effen1337** [comment]
@yeldarby I used @dteh's fork and I actually get `Error [TypeError]: Canvas expected` when I use a worker.

**Comment by dteh** [comment]
@effen1337 perhaps it is a platform issue? I've successfully built and deployed it on OSX 10.15.6 and CentOS 7.8.2003

**Comment by yeldarby** [comment]
I'm running it in Docker (Ubuntu 18.04 base) with no problems so far.

**Comment by yeldarby** [comment]
Hmm, I may have spoken too soon. I'm getting weird errors now too

```
TypeError: Method Image.SetSource called on incompatible receiver
    at setSource (transform/node_modules/canvas/lib/image.js:91:13)
    at Image.set (transform/node_modules/canvas/lib/image.js:65:7)
    at transform/node_modules/canvas/index.js:34:15
```

**Comment by effen1337** [comment]
After checking out this https://github.com/microsoft/node-pty/issues/405#issuecomment-622453030

Seems like even `NAN_MODULE_WORKER_ENABLED` doesn't get the job done on node.js 10 and up

**Comment by nickpharrison** [comment]
Is there any chance that the fix in @dteh's fork could be implemented in the main repo?

**Comment by nickpharrison** [comment]
I've been able to replicate the "Canvas expected" error @yeldarby mentioned above (on Windows) after building @dteh's fork. It occurs when calling `canvas.createCanvas` from the main thread only after the same function has previously been called from a worker thread.

**Comment by afbarbaro** [comment]
Any chance that this fix will be merged into the main repo soon? Would love to be able to use worker threads for cases where we're working with multiple canvases so each one can be operating on it's own worker thread without blocking the main thread.

**Comment by LinusU** [comment]
> Any chance that this fix will be merged into the main repo soon?

I must have missed that PR, would you mind pointing me to it? 🤔

**Comment by zbjornson** [comment]
#1734 was the PR, but there's more work required than what's been suggested above. I think the font registrations and saved canvas context states need per-context storage instead of static storage, might be more.

**Comment by niyan-ly** [comment]
Node.js worker will share memory. 
Native addon may run in different context, but share the same memory. 
The `canvas expected` error raise because class `Canvas` create a new constructor template `ctor` each time the addon was loaded, and bind this template to its static member `constructor`. Then class `Context2d` try to find its instance on `New`. But since workers are sharing memory, only worker who load `canvas.node` lastly will work as expected(FILO likely).

And that's what `context-aware` means, workers runs in different context, but share memory.

![image](https://user-images.githubusercontent.com/28030655/151531877-ea55d399-1f04-4c73-a2f1-9609a1472ce4.png)

**Comment by LosTigeros** [comment]
If you're using it on a backend I'm using alternative way.
Basically I'm using BullMQ queue with Sandboxed processes. It works like a charm and it was what I exactly needed.

**Comment by SonahtQ** [comment]
Any progress on that? or working workaround?

**Comment by SBRK** [comment]
I too am bothered by this limitation.

**Comment by kylerchin** [comment]
Please fix this  😭 we all need this

**Comment by DrMeepso** [comment]
Any updates?

**Comment by Balearica** [comment]
@zbjornson It looks like this issue was accidentally closed as completed due to the following sentence in #2235: `We're closer to being able to close #1394`.  It looks like the issue is still outstanding and should be re-opened.

**Comment by BowgartField** [comment]
any updates ?

**Comment by qbqaf** [comment]
Also just ran into the issue of `Module did not self-register`, even in `3.0.0-rc2`.

Would love to see this happen, would really make my time a lot easier ;D

**Comment by deflexable** [comment]
any updates on this?

**Comment by Zirafnik** [comment]
As @anshul-kai pointed out above, the current workaround is to use Node.js `child_processes` instead of `worker_threads`.

From what I've tested, that way works fine.

**Comment by lvcabral** [comment]
The proposed `workaround` is good for all use cases, I need it as a Worker, because my application uses SharedArrayBuffer to communicate to the Worker, there are other issues, that I would have to rewrite several portions of the app. The ideal was really to have Canvas support in Worker, just like the Browser has.

**Comment by chearon** [comment]
I'm more than halfway through redoing the font stack and prebuilds, and this is on my radar. I don't think we will have any more global/static data after that. It will definitely be done, it just depends on how much free time I have.

**Comment by lvcabral** [comment]
@chearon just a quick check, any update for us?

**Comment by chearon** [comment]
Yes, I finished all the hard stuff for v4 (#2476)  and it will be out soon! I just tried 4 canvases in parallel via worker threads. I had to tweak a few things, but it worked, with different fonts and SVG rendering in parallel:

<img width="200" height="200" alt="Image" src="https://github.com/user-attachments/assets/5a10b786-5a39-4ca8-9779-710b0d291214" />
<img width="200" height="200" alt="Image" src="https://github.com/user-attachments/assets/b551fe70-2b50-46f3-bb22-b5895a330f42" />
<img width="200" height="200" alt="Image" src="https://github.com/user-attachments/assets/5232fbe4-1ac5-4903-afd1-5d0bc070f62c" />
<img width="200" height="200" alt="Image" src="https://github.com/user-attachments/assets/d949b17b-3386-4114-a860-4084de45e8c5" />

Development is happening on the `next` branch. I have to review and test more, then figure out the new way to distribute binaries (#2595) but we're very close.

**Comment by lvcabral** [comment]
@chearon I really appreciate the effort! This will unlock the CLI version of my open source Roku simulator to also support Task threads!
